### ISSUE TITLE
review/update design to support "confirmation and completed" workflow

### ISSUE THREAD

**Issue by monkeypants** [gs-gs/parks-bce-integration-bookable-things#13]
Currently, the API allows the reservation to be updated using PATCH as described in the following  "Reservation update (current)". 

BCE has an interesting data payload associated with "confirmation" and "completed" events. We probably do want this data, but what is the best way to handle it? The "Reservation update (options)" section discusses how we might handle this interesting data, the purpose of this ticket is to agree on an approach and then implement it.

## Reservation update (current)
`PATCH /reservations/{reservation_id}/`
Request:
```
{"field1": "value1", ...}
```
Validations are applied.

Some common use-cases:

* delivery org: accept reservation - update status to “accepted”
* delivery org: deny reservation - update status to “denied” (with some note probably)
* delivery org: finaize booking after fulfillment (status=”completed”)
* agent: request reservation cancellation (status=”cancellation_requested”)
* delivery_org: confirm reservation cancellation (status=”cancelled”)

## Reservation update (options)

1. edit the schema in place - when the new data is available, patch-up the reservation (using the same schema)
2. update sub-schema. E.g. if we have a "confirmed" and "confirmation_details" attributes, with some validation logic saying "if confirmed then confirmation_details must not be none, and if not confirmed then confirmation_details must be none, And the same for "completed"/"completion_details".
3. we could have different endpoints:
  * `POST {reservation_details} /reservations` -> reservaton_id
  * `POST {confirmation_details} /reservations/{id}/confirmations`
  * `POST {finalization_details} /reservations/{id}/finalization`

The current implementation is as-per **1.**. The problem with this is that we loose the information about difference between initial booking, confirmed numbers and actual/finalised numbers.

The second option (**2.**) is a bigger, more complicated schema, but is also compatible with the current implementation. However we probably do need to redesign the schema and work through the details to make sure there are no problems that come out of that.

The third option (**3.**) could work, is it really more user-friendly? Are these state changes universal or BCE specific? Will we always have a schema/extra_data json attribute associated with confirmation and finalization events?

**Comment by dormerod** [comment]
@ltankey says:

(doesn't consider authorisation)

There's a thing, and there's a schema behind it and in that schema there are multiple branches of subschema.

You patch the form payload.

This is options 1 and 2, using Django revisions.

---

@monkeypants says:

In the current API spec it's PATCH, but we should PUT an updated version of the reservation details (extra_data).

Clarification: it's a PATCH on the Django model, a PUT on the form.

This means options 1 and 2 are the same from @ltankey's perspective.

**Comment by dormerod** [comment]
@koriaf and @mr-love can you please take a look at this and see if you agree with the comment above?

If so, can you please make an example for the docs @mr-love and ask @ltankey to review it?

**Comment by koriaf** [comment]
There is `extra_data` parameter which accepts JSON of any format (top level assumed to be dict). If we are talking about something organisation-specific then it will be the best place to hold any extra information (like text reason fields and numbers of passengers registered/attended/refunded/happy/etc).

That means that server-side validation of such things is not implemented and I'll need exact set of rules to do that (apart of these examples with confirmed reason).

And please note that product has `extra_data_public` field, which theoretically can hold some validation rules for these fields - if we decide how they look like.

**Comment by mr-love** [comment]
So if I'm understanding correctly the 'sub-schemas' for confirmations and finalisation are the same as the base schema? What is the use case for keeping the data at each status point? I would be leaning towards just using revisions otherwise.

**Comment by monkeypants** [comment]
I fear I may have over-complicated this.

If we are talking about the contents of the extra_data attribute, then we need to think about sub-schema validation and so on. It will be different logic for different types of product that has been booked, will require schema-versioning and linking schema-versions to bookable Products (etc). We don't currently have that apparatus or any user-stories that need it.

But I think we only need to solve for two situations now. Is this right?

1. Confirmation event (with data).
2. Completion event (with data, may be some days post-facto).

In part this could be solved with `PATCH {status=confirmed} /reservations/{id}` and `PATCH {status=completed} /reservations/{id}`. That would presumably update `Reservation.status` (without any need to touch `Reservation.extra_data`), and could raise errors for simple state machine violations and permission violations.

The "send additional info" business is harder. We have been talking about updating `Reservation.extra_data` but that's where we need to conform with future schema changes without breaking BCE's stuff - so it's hard to get right. 

Here's my suggestion for discussion: What if we left extra_data for schema-compliant form payloads, and made a new object  similar to `ReservationNote`, something like `ReservationData`. Instead of a text field, it could have a "payload_type" field and two JSON fields ("payload" and "schema"). The API could be something like:
*  `POST {payload+schema} /reservations/{id}/confirmation_data`; and
*  `POST {payload+schema} /reservations/{id}/completion_data`.

These POSTs could journal the data and schema as new ReservationData objects (of appropriate type). If BCE started using a different schema in the future, they could tell us after the fact ("oh, by the way, we stopped sending that data we used to send and started sending some new format" "yes, we notices that our scripts stopped working"). We couldn't break the contract that way, only BCE could (please don't make breaking schema changes without talking to us first, otherwise we might have some difficulty making use of the data you send us - but if you get hit by a bus and we really wanted to we could probably figure it out since you are posting the schema with every payload).

Furthermore, we could make an implicit `PATCH {status=confirmed} /reservations/{id}` on a valid POST to the `confirmation_data`, and an implicit `PATCH {status=completed} /reservations/{id}` on a valid POST to the `completion_data`. So the error messages on the POST would be the same as for PATCH plus another one "that payload does not match the supplied schema" (permissions would be aligned).

**Comment by monkeypants** [comment]
in case it wasn't obvious - this leaves the `extra_data` JSON as some organisation-specific stuff (actually Product specific, i.e. corresponding to the form that was filled out when the booking was made, "would you like fries with that" or "are those gherkins allergic to moonlight" etc).

**Comment by koriaf** [comment]
Although the described approach increases complexity, I think it's fine and will offer the client a way to save whatever they want.
If we want to keep things simper we can avoid creating any new endpoints and objects and still use `extra_data` - and if I say that current form data is always saved as a subkey `formData` which leaves subkeys `confirmationData` and `completitionData` available it could simplify things a lot - removing problems which you are trying to solve. This way the form schema won't look at the data aprt of `formData` and any other (including future) schemas will just validate the keys they are responsible to.

There is possible issue about access (extra_data is visible to all related parties) but I don't remember any requirements here so it should be fine.

keen to hear BCE opinion about it. As always, I tend to pick the most universal solution - allowing users to do whatever they want without affecting other users and asking us to tune things specially for them.

![Screenshot from 2021-02-17 10-37-16](https://user-images.githubusercontent.com/131850/108185022-2210ba80-710c-11eb-8bf6-fcb86f22048c.png)

**Comment by monkeypants** [comment]
OK, so it doesn't matter if we save the ReservationData in the one JSON blob or as it's own entity. That's an implementation detail. The blocking question is on the API design. In summary:

* `PATCH {status=confirmed} /reservations/{id}`
* `POST {payload+schema} /reservations/{id}/confirmation_data` (optional, implies PATCH if not already done)
* `PATCH {status=completed} /reservations/{id}`
* `POST {payload+schema} /reservations/{id}/completion_data` (optional, implies PATCH if not already done)

With errors based on state machine and permission violations, plus validation error (payload is not valid per schema submitted it claims to conform with). The state machine on `confirmed -> confirmed` and `finalised -> finalised` is not an error, it's OK to do that (finalisation and confirmation are idempotent).

If @simon-tma agrees, let's do that.

**Comment by monkeypants** [comment]
note:
* API user can `POST ...` as many times as they like. We keep all versions (in a journal) but surface the most recent as required. So it's kind of like a journaling PUT.

That means you can send an initial confirmation, and then an updated one (with e.g. "5 day out covid-19 policy compliance declaration") at any time.

**Comment by monkeypants** [comment]
Another option - reasonable equivalent - is to allow `PUT {payload+schema} /reservations/{id}/confirmation_data` (and equivalently `.../completion_data`)  rather than POST. It's not much different from our side but POST seems aesthetically better because PUT implies a GET should also be possible (and there isn't one). I don't think it matters, if people prefer PUT rather than POST I'll be happy to go with the majority.

**Comment by simon-tma** [comment]
Looks good to me. We'll need to figure out the schema structure, but that's separate from the implementation.

**Comment by monkeypants** [comment]
IIRC we provided you a hand-rolled JSON schema for the initial reservation, and that makes sense because it has to meet our reservation management needs. It's probably better that you provide us the schema for your `confirmation_data` and `completion_data`, since we "will take what we can get" from you in that area. Main thing is to settle on something and keep it constant, or at least let us know when you have more-or-less stopped changing it, so we can then start to look at how we deal with whatever you are sending us.

**Comment by koriaf** [comment]
Who is sending confirmation data and completion data? Delivery org or agent org or both?
I assume that only delivery org can change status to "confirmed" but it's not clear if agent can do something and what exactly

**Comment by koriaf** [comment]
Documentation is here https://github.com/gs-gs/parks-bce-integration-bookable-things/commit/26c1e2ceff7ec5e5117db7b185631c41c053d619 and the code is deployed to int.

Please note we don't do so many validations here on that point - will be implemented once requested.

**Comment by monkeypants** [comment]
> Who is sending confirmation data and completion data

good question. Agree it's the delivery org that does "confirmed".

With BCE integration it's a bit funny because they have both tokens. But an Agent can't confirm their own booking, it has to be the delivery org. Agent asks (booking request?), delivery org answers (confirmed!).

> ...and completion data

The most obvious answer is that this is delivery org too, but the obvious answer is probably wrong. BCE is wearing their Agent hat when they send "here is some more information about the booking that we are sending post-facto". So I think both Agent and Delivery Org can send this.

in summary, I'm agreeing with your as-documented behaviour :)

**Comment by krylataya** [comment]
@koriaf all works fine but I'm a bit misleaded by fact that confirmation_data and completion_data fields are described as required in documentation but in reality it's optional

**Comment by monkeypants** [comment]
Apart from Julia's comment about documentation, is there anything left to do on this ticket?

**Comment by koriaf** [comment]
Can be closed if client is happy
